cirq-core/cirq/sim/simulator.py

```diff
@@ -57,7 +58,7 @@
 
 TStepResult = TypeVar('TStepResult', bound='StepResult')
 TSimulationTrialResult = TypeVar('TSimulationTrialResult', bound='SimulationTrialResult')
-TSimulatorState = TypeVar('TSimulatorState')
+TSimulatorState = TypeVar('TSimulatorState', bound=Any)
```

Comment:
What does this bound do? I would think since everything is an `Any` then this would be equivalent to having no bound at all.

Reply:
There was a place where `Optional[TSimulatorState]` caused a mypy error without this. I'll re-check before resolving this.

Reply:
Okay, it complains below where we do `_final_simulator_state.qubit_map` that `Item "object" of "Union[Any, TSimulatorState]" has no attribute "qubit_map"`. Prior to this PR, `_final_simulator_state` was just type `Any`. I guess the default generic bound is `object`, not `Any`.

Anyway it seems like a fair complaint. We should have an interface bound for TSimulatorState that includes `qubit_map`. Added issue #5270 to track.